### DISCUSSION TITLE
Component Caching

### DISCUSSION THREAD

**Discussion by regisphilibert** [withastro/roadmap#733] [Proposal]
# Summary

A way for Astro to cache the HTML output of a component so it only needs to be processed once or a limited time during the build rather than on every page/route. For all subsequent invocation, the cached HTML will be inserted in the page.

This feature is available on the Hugo framework see https://gohugo.io/functions/partials/includecached/

# Background & Motivation

There has been a lot of discussion and feature requests about making data persistant across the build (https://github.com/withastro/roadmap/discussions/670, https://github.com/withastro/roadmap/discussions/603) or file caching (https://github.com/withastro/roadmap/discussions/612). Both solution would be great for limiting API calls, or simply download a file that can be useful through the builds and across routes and components.

But I feel a lot of motivation behind those two requested features is mainly avoiding expansive logic repetition. Some needs to happen for every page.

Being able to « cache » the HTML output of a component, so that Astro will, on its next use simply inject the cached HTML rather than process the component logic would solve a lot of the problems involved with persistant data or file caching.

# Goals

- Enable users to limit repeated processing of highly static yet sometime heavy logic components (navigation, footer, author bio etc…).
- Enable users to set via a directive and its optional cacheKey a way to limit said caching to user define conditions.

# Example

The caching could be invoked via a new template directive (`is:cached={string}`)who could take a value for a cacheKey or cache all the time if the later is absent.

```astro
<Header is:cached={section} />
<slot/>
<Footer is:cached />
```

**Comment by lilnasy** [comment]
In my experience, the rendering performance itself has been really good - pure pages render in less than 5ms. I think the concern might be `fetch()` and `fs.readFile()`. If that is the case, would it be possible for your use case to cache into an in-memory map and share it across pages?

**Reply by regisphilibert** [reply]
I would love to find a solution for the problem you described, but could not, browsing the various threads and discussions (some listed here). Hence me suggesting a global solution available to everyone and which would remove the need to solve the ones I discussed and probably other use case I did not think about.

Though I'd be curious if you have some pointers for the "in-memory" solution you describe? The term is not familiar to me.

**Reply by lilnasy** [reply]
I think it's best described via a code example:
```js
//  src/cachedFetch.js
const map = new Map

// Makes an http request when called the first time.
// Reuses data from map when called with the same url string a second time.
export default async function (url) {
    const cachedData = map.get(url)
    if (cachedData) return cachedData
    const response = await fetch(url)
    const freshData = await response.json()
    map.set(url, freshData)
    return freshData
}
```
```astro
---
// src/pages/index.astro
import fetch from "../cachedFetch.js"
const data = await fetch(endpoint)
---
```

**Reply by regisphilibert** [reply]
This has been a morning of revelations... I found about `global` and tried using it with success, but your suggestion is cleaner and more predictable across hosts I suppose... I'm surprised I never read about this, but I guess it's because it is so obvious. Thanks!

**Reply by wassfila** [reply]
@lilnasy components caching is unfortunately not only about fetching but also about Astro logic build time, I'll argue with my example on a separate comment where I have a solution for this.

**Comment by wassfila** [comment]
this is so interesting, I just had an idea Today, I implemented it and only once I was done, I though of searching in roadmap and I found this.

# Rationale
Caching Astro components can be a life saver in some scenarios and a scalability enabler. I faced an issue with a big website where I had a menu hat could link to all pages. I know pretty well that designers purists would flag that as a design mistake, but that might not be always up to me, plus the line on when it becomes too much is subjective, a margin within which there's a lot of room to improve Astro's build performance. The same menu Layout in all pages is just one example, I could think of many others that would have a repetitive widget in all pages and so on. My first naive solution was to shift menu injection to the client, and it worked, I generate the menu only once in json and let the client inject it. That solution is so unpractical compared to the possibility of simply caching an Astro component.
Preventing menu rendering on every page has improved the build performance as follows :
* Page build time down to ~ 100 ms or so from > 4 sec !!!!!!!
* Full site build time down to ~ 3 min from ~ 90 min !!!!!!!

# Solution
I do have a solution that performs this without requiring Astro's internal magic, the trick is based on `Astro.slots.render()` https://docs.astro.build/en/reference/api-reference/#astroslotsrender

here a link to a functional small example https://github.com/MicroWebStacks/astro-examples/tree/main#38_cached-component

and here an extract from the readme in this discussion for convenience
usage is as simple as this
```js
    <Cache>
        <Menu />
    </Cache>
```
can also have different names that act as store keys

```js
    <Cache name="section1">
        <Menu section="section1"/>
    </Cache>
```

Note1 : This will cache all calls from all isntances, if the intended usage is a different cache set for every astro component file, then it is still required that the user gives a different `name` for every instance, e.g. prefix component path (with e.g. a [vite __filedir plugin](https://github.com/MicroWebStacks/astro-big-doc/blob/97412f4914f7a8f821dc762f1023aed62091ae75/src/libs/vite-plugin-filedir.js#L5) ) and **this is a point where Astro's magic could help simplify for the users**

Note2: This is caching production only for convenience during development, if caching is also intended during development (not recommended) then this line in `Cache.astro` can be changed to remove the PROD condition
```js
if(cache_has(name) && import.meta.env.PROD){
```

Test, see build log, both pages build are logged while the rendering of the cached component is logged only once
```
index> rendering index page
 ** rendering ** 'default' => cache_set()
page2> rendering page2
```

**Comment by jamesli2021** [comment]
@wassfila When you set the `<Cache name="abc">...</Cache>`
I could only see it set as `undefined` value but set with proper data when you don't set a name.

Using `slot` should works
```
    <Cache name="section1">
        <Menu slot="section1"/>
    </Cache>
```

**Comment by wildfiremedia** [comment]
@wassfila What do you think of this idea?
https://github.com/withastro/roadmap/discussions/1118

@regisphilibert I think my idea might or might not be a good design but at least, it's easy to add directive in a `<Fragment is:static>`.

**Comment by delucis** [comment]
In case it’s helpful, here’s an example I made of a simple `<Cacher>` component that only renders its children on the first page render and then reuses the HTML: https://stackblitz.com/edit/github-qtj7yw66?file=src%2Fcomponents%2FCacher.astro

Obviously the implementation here is a toy version — an in-memory `Map` containing cached HTML. In reality you’d replace it with something else probably (whether an in-memory LRU cache or a Redis instance or whatever) and have a way to invalidate the cache probably.

**Reply by wildfiremedia** [reply]
Yeah, I’ve been using @wassfila's Maps caching solution for a while.
